sphinx/writers/latex.py

```diff
@@ -1408,45 +1413,47 @@ def visit_entry(self, node):
         cell = self.table.cell()
         context = ''
         if cell.width > 1:
-            self.body.append('\\multicolumn{%d}' % cell.width)
             if self.table.col == 0:
-                self.body.append('{|l|}{\\relax ')
+                self.body.append('\\sphinxstartmulticolumn{%d}{\\vline}%%\n' % cell.width)
             else:
-                self.body.append('{l|}{\\relax ')
-            context += '\\unskip}\\relax '
+                self.body.append('\\sphinxstartmulticolumn{%d}{}%%\n' % cell.width)
+            context = '\\sphinxstopmulticolumn\n'
         if cell.height > 1:
-            self.body.append('\\multirow{%d}{*}{\\relax ' % cell.height)
-            context += '\\unskip}\\relax '
-        if (('morecols' in node or 'morerows' in node) and
-           (len(node) > 2 or len(node.astext().split('\n')) > 2)):
-            self.in_merged_cell = 1
-            self.literal_whitespace += 1
-            self.body.append('\\eqparbox{%d}{\\vspace{.5\\baselineskip}\n' % id(node))
-            self.pushbody([])
-            context += '}'
+            # \sphinxmultirow 2nd arg is index in a LaTeX array storing contents height
+            self.body.append('\\sphinxmultirow{%d}{%d}{%%\n' % (cell.height, cell.cell_id))
+            self.body.append('\\begin{varwidth}[t]{\\sphinxcolwidth{%d}{%d}}'
+                             '\\strut\n' % (cell.width, self.table.colcount))
+            context = ('\n\\par\\vskip-\\baselineskip\\strut'
+                       '\\end{varwidth}%\n}%\n') + context
+        elif cell.width > 1:
+            self.body.append('\\begin{varwidth}[t]{\sphinxcolwidth{%d}{%d}}'
+                             '\\strut\n' % (cell.width, self.table.colcount))
+            context = ('\n\\par\\vskip-\\baselineskip\\strut'
+                       '\\end{varwidth}\\par\\strut\n') + context
+        if cell.width > 1 or cell.height > 1 or \
+           (len(node) > 2 and len(node.astext().split('\n')) > 2):
+            self.needs_linetrimming = 1
         if isinstance(node.parent.parent, nodes.thead):
             if len(node) == 1 and isinstance(node[0], nodes.paragraph) and node.astext() == '':
                 pass
             else:
                 self.body.append('\\sphinxstylethead{\\relax ')
-                context += '\\unskip}\\relax '
-        if len(node.traverse(nodes.paragraph)) >= 2:
-            self.table.has_problematic = True
```

Comment:
@tk0miya because of removal I did of this setting of ``has_problematic``, in some cases the colspec gets set to `|l|...|l|` rather than using ``\X`` specifier. As a result a cell containing  a long paragraph is set horizontally and text will flow in right margin.

Currently thinking of general solution of handling long cell contents. There is a column specifier `V` provided by package `varwidth`. Maybe that could be used generally speaking and `tabulary` package could be dropped. I need to investigate more. My task is not finished here...